### ISSUE TITLE
Naked Rationale

### ISSUE THREAD

**Issue by brentvee** [NakedObjectsGroup/NakedObjectsFramework#6]
Hey Richard,

Hope you're well.
I'm hoping you can educate me and let me peek into your mind for a short while.
What was your rationale behind having NOF be so 'naked'?

By this I mean, the opt-out strategy of the framework and the need for a developer to hide things.

Let me play devils advocate here.
Our domain is exposed through several different endpoints, of which only one is a NOF website.
It would be quite nice if a developer could rather specifically say, "yes i want this particular method exposed on the UI, and here are its complementary methods etc".
Same would go for entities. 
If an entity was missing, lets say a `Navigable` attribute, then it would present as a label on the UI and not a link.

With this opt-in strategy a developer could add several `public` methods into the domain and entities without unknowingly or accidentally exposing something onto the UI (not the ideal scenario, since the dev should know what they're doing).
It would also allow new developers to add value to the domain without the overhead of NOF knowledge.

`[Eagerly(EagerlyAttribute.Do.Rendering)]` awaiting your thoughts.

Again, many thanks for a great framework that I really enjoy using.

**Comment by richardpawson** [comment]
Brent

Sorry for the delay  -  I was on vacation.

When we created the first version of NOF, we debated the opt-in vs. opt-out patterns.  It could have worked perfectly well either way.  We chose the opt-out pattern, in part to encourage developers to think of everything being exposed by default - and therefore to make their models better (like the idea of having a restaurant kitchen visible to the diners  -  it, arguably, encourages better cooking practices! Or the architecture of the Beaubourg in Paris, for that matter). Bear in mind we were trying to counter the predominant pattern of keeping entities hidden from the user and exposed only via dedicated View Models or user views.

It would not be very difficult to change the frarmework to an opt-in pattern instead  -  but we have existing users/applications to think of.  And I would not recommend doing it yourself unless you want to manage your own fork of the framework and integrate all the upgrades each time.

I'm not sure I buy the comment that the opt-in pattent, would also allow new developers to add value to the domain without the overhead of NOF knowledge.' though.  The overhead of NOF knowledge required is very small compared to most frameworks.  And the idea of a developer 'adding value' to domain code with no impact on the users is not, IMO, a realistic or desirable idea.
